modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxWithKeyContentionSelfTest.java

```diff
@@ -0,0 +1,329 @@
+import org.apache.ignite.testframework.MvccFeatureChecker;
+import org.apache.ignite.testframework.junits.WithSystemProperty;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionIsolation;
+import org.junit.Test;
+
+import java.util.concurrent.CountDownLatch;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_DUMP_TX_COLLISIONS_INTERVAL;
+import static org.apache.ignite.transactions.TransactionConcurrency.OPTIMISTIC;
+import static org.apache.ignite.transactions.TransactionConcurrency.PESSIMISTIC;
+import static org.apache.ignite.transactions.TransactionIsolation.READ_COMMITTED;
+import static org.apache.ignite.transactions.TransactionIsolation.REPEATABLE_READ;
+
+/** */
+public class TxWithKeyContentionSelfTest extends GridCommonAbstractTest {
+    /** Client flag. */
+    private boolean client;
+
+    /** Near cache flag. */
+    private boolean nearCache;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String name) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(name);
+
+        cfg.setConsistentId("NODE_" + name.substring(name.length() - 1));
+
+        if (client)
+            cfg.setClientMode(true);
+
+        cfg.setDataStorageConfiguration(
+            new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(
+                    new DataRegionConfiguration()
+                        .setMaxSize(20 * 1024 * 1024)
+                )
+        );
+
+        TestRecordingCommunicationSpi commSpi = new TestRecordingCommunicationSpi();
+
+        cfg.setCommunicationSpi(commSpi);
+
+        cfg.setCacheConfiguration(getCacheConfiguration(DEFAULT_CACHE_NAME));
+
+        if (client) {
+            cfg.setConsistentId("Client");
+
+            cfg.setClientMode(client);
+        }
+
+        return cfg;
+    }
+
+    /** */
+    protected CacheConfiguration<?, ?> getCacheConfiguration(String name) {
+        CacheConfiguration<Object, Object> ccfg = new CacheConfiguration<>(name)
+            .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+            .setWriteSynchronizationMode(CacheWriteSynchronizationMode.FULL_SYNC)
+            .setAffinity(new RendezvousAffinityFunction(false, 16))
+            .setBackups(2)
+            .setStatisticsEnabled(true);
+
+        if (nearCache)
+            ccfg.setNearConfiguration(new NearCacheConfiguration<>());
+
+        return ccfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * @throws Exception If failed.
```

Comment:
Add description.

Reply:
all description done in calling func.

Reply:
cargo cult, all helpful description given in calling method, i don`t want to add copy paste here.

Reply:
I'll leave it up to the next reviewer.

Reply:
I believe that brief comments are useful here, cause they will make an intention of a test more explicitly.